android-voice-processor/src/main/java/ai/picovoice/android/voiceprocessor/VoiceProcessor.java

```
@@ -0,0 +1,305 @@
+            return;
+        }
+
+        isStopRequested.set(true);
+        try {
+            readThread.get();
+            readThread = null;
+        } catch (ExecutionException | InterruptedException e) {
+            throw new VoiceProcessorException(
+                    "An error was encountered while requesting to stop the audio recording",
+                    e);
+        } finally {
+            isStopRequested.set(false);
+        }
+    }
+
+    @SuppressLint("MissingPermission")
+    private void read() {
+        final int minBufferSize = AudioRecord.getMinBufferSize(
+                this.sampleRate,
+                AudioFormat.CHANNEL_IN_MONO,
+                AudioFormat.ENCODING_PCM_16BIT);
+        final int bufferSize = Math.max(this.sampleRate / 2, minBufferSize);
+
+        AudioRecord recorder;
+        try {
+            recorder = new AudioRecord(
+                    MediaRecorder.AudioSource.MIC,
+                    this.sampleRate,
+                    AudioFormat.CHANNEL_IN_MONO,
+                    AudioFormat.ENCODING_PCM_16BIT,
+                    bufferSize);
+        } catch (IllegalArgumentException e) {
+            onError(new VoiceProcessorException(
+                    "Unable to initialize audio recorder with required parameters",
+                    e));
+            return;
+        }
+
+        if (recorder.getState() != AudioRecord.STATE_INITIALIZED) {
+            onError(new VoiceProcessorException(
+                    "Audio recorder did not initialize successfully. " +
+                            "Ensure you have acquired permission to record audio from the user."));
+            return;
+        }
+
+        final short[] buffer = new short[this.frameLength];
+        try {
+            recorder.startRecording();
+
+            while (!isStopRequested.get()) {
+                if (recorder.read(buffer, 0, buffer.length) == buffer.length) {
+                    onBuffer(buffer);
+                }
```

This is the same scenario we had in `ios-voice-processor` where the the maximum buffer that the recorder gave was less than the requested (for Eagle) and then it silently dropped frames. We should strongly consider either a) internally buffering in here in case the call to `.read()` comes up short, or b) finding out what the maximum is and guarding against it as user input.

---

I went with putting an event into the `errorListener` because it's also possible that the recorder isn't giving any bytes or more than the frame. In that case they need to correct their code.